python/ray/autoscaler/_private/autoscaler.py

```diff
@@ -266,24 +283,73 @@ def _sort_based_on_last_used(self, nodes: List[NodeID],
                                  last_used: Dict[str, float]) -> List[NodeID]:
         """Sort the nodes based on the last time they were used.
 
-        The first item in the return list is the least recently used.
+        The first item in the return list is the most recently used.
         """
         updated_last_used = copy.deepcopy(last_used)
-        now = time.time()
+        # Add the unconnected nodes as the least recently used (the end of
+        # list). This prioritizes connected nodes.
+        least_recently_used = -1
         for node_id in nodes:
             node_ip = self.provider.internal_ip(node_id)
             if node_ip not in updated_last_used:
-                updated_last_used[node_ip] = now
+                updated_last_used[node_ip] = least_recently_used
 
         def last_time_used(node_id: NodeID):
             node_ip = self.provider.internal_ip(node_id)
             return updated_last_used[node_ip]
 
         return sorted(nodes, key=last_time_used, reverse=True)
 
-    def _keep_min_worker_of_node_type(self, node_id: NodeID,
-                                      node_type_counts: Dict[NodeType, int]):
-        """Returns if workers of node_type should be terminated.
+    def _get_nodes_allowed_to_terminate(
+            self, sorted_node_ids: List[NodeID]) -> Dict[NodeID, bool]:
+        """Returns the nodes allowed to terminate for request_resources().
+
+        Args:
+            sorted_node_ids: the node ids sorted based on last used (LRU last).
+
+        Returns:
+            nodes_allowed_to_terminate: whether the node id is allowed to
+                terminate or not.
+        """
+        nodes_allowed_to_terminate: Dict[NodeID, bool] = {}
+        head_node_resources: ResourceDict = copy.deepcopy(
+            self.available_node_types[self.config["head_node_type"]][
+                "resources"])
+        max_node_resources: List[ResourceDict] = [head_node_resources]
+        resource_demand_vector_worker_node_ids = []
+        # Get max resources on all the non terminated nodes.
+        for node_id in sorted_node_ids:
+            tags = self.provider.node_tags(node_id)
+            if TAG_RAY_USER_NODE_TYPE in tags:
+                node_type = tags[TAG_RAY_USER_NODE_TYPE]
+                max_node_resources.append(
+                    copy.deepcopy(
+                        self.available_node_types[node_type]["resources"]))
+                resource_demand_vector_worker_node_ids.append(node_id)
+        # Since it is sorted based on last used, we "keep" nodes that are
+        # most recently used when we binpack. We assume get_bin_pack_residual
+        # is following the given order here.
+        used_resource_requests: List[ResourceDict]
+        _, used_resource_requests = \
+            get_bin_pack_residual(max_node_resources,
+                                  self.resource_demand_vector)
```

Comment:
This exact same logic to calculate adjusted min workers is duplicated in resource_demand_scheduler, can we unduplicate it?

Reply:
Ideally,
- resource_demand_scheduler has no concept of request_resources
- we calculate adjusted_min_workers: Dict[NodeType, int] here, and pass it into resource_demand_scheduler when getting nodes to launch

That deduplicates the logic (that calls bin packing for increasing min workers for request_resources())

Reply:
Can we make this a TODO to address later? It is not exactly duplicated logic and to make it work it would require some not straightforward refactor.
We also don’t really have exactly a duplicate logic because in the resource_demand_scheduler we count them, but outside we don’t. The only duplicate is the binpacking. Which is also not exactly duplicate, we binpack internally on all nodes (pending/launching/running) but outside we binpack on (pending/running), the logic differs between terminating and launching. When launching we are very conservative in adding nodes, when terminating we are very conservative in terminating nodes (opposite of conservative in adding). If a node is launching in upscaling, we count it, if it is launching when we terminate we don't count it (potentially keeping non-pending nodes for longer time).

Reply:
Fair enough given the urgency of the issue. Please leave a TODO comment here.

Reply:
I added a TODO at the beginning of the function (line 305). 
Thanks!